catalyst/core/callbacks/checkpoint.py

```diff
@@ -251,37 +327,65 @@ def on_stage_start(self, state: State):
             _load_checkpoint(filename=self.resume, state=state)
             self.resume = None
 
-    def on_epoch_end(self, state: State):
-        """@TODO: Docs. Contribution is welcome."""
+    def on_epoch_end(self, state: State) -> None:
+        """
+        Collect and save checkpoint after epoch.
+
+        Args:
+            state (State): training state
+        """
         if state.stage_name.startswith("infer") or state.is_distributed_worker:
             return
 
-        checkpoint = _pack_state(state)
-        self.process_checkpoint(
-            logdir=state.logdir,
-            checkpoint=checkpoint,
-            is_best=state.is_best_valid,
-            main_metric=state.main_metric,
-            minimize_metric=state.minimize_metric,
-        )
+        if self.save_n_best > 0:
+            checkpoint = _pack_state(state)
+            self.process_checkpoint(
+                logdir=state.logdir,
+                checkpoint=checkpoint,
+                is_best=state.is_best_valid,
+                main_metric=state.main_metric,
+                minimize_metric=state.minimize_metric,
+            )
 
-    def on_stage_end(self, state: State):
-        """@TODO: Docs. Contribution is welcome."""
+    def on_stage_end(self, state: State) -> None:
+        """
+        Show information about best checkpoints during the stage.
+        Load model specified in ``load_on_stage_end``.
+
+        Args:
+            state (State): training state
+        """
         if state.stage_name.startswith("infer") or state.is_distributed_worker:
             return
+        # store latest state
+        if self.save_n_best == 0:
+            # full checkpoint
+            checkpoint = _pack_state(state)
+            self._save_checkpoint(
+                logdir=state.logdir,
+                checkpoint=checkpoint,
+                suffix="last",
+                is_best=True,  # will duplicate current (last) as best
+                is_last=False,  # current state is last state
```

Comment:
do you think we really need `best`/`last`checkpoints duplication?

Reply:
Yeah, we need to save a current state (checkpoint on stage end) as last and best states.
With the current implementation of `ConfigExperiment` best state is required. This state used for initializing model weighs at stage start:

* https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/experiment/config.py#L162-L169
* https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/experiment/config.py#L280-L302

Reply:
Deal, let's make it compatible with current Experiment logic :)
Experiment-Runner logic refactoring - is for another pull request